src/common/checksum.c

```diff
@@ -20,218 +20,509 @@
-		return -DER_INVAL;
-	}
-	memcpy(csum_buf->cs_csum, csum->dc_buf,
-	       csum_dict[csum->dc_csum].cs_size);
+	D_ALLOC(ctx, sizeof(struct mh_sha1_ctx));
+	obj->dcs_ctx = ctx;
 
-	return 0;
-#else
-	return mchecksum_get(csum, csum_buf->cs_csum, csum_buf->cs_buf_len,
-			     MCHECKSUM_FINALIZE);
-#endif
+	return mh_sha1_init(ctx);
 }
-inline int
-daos_csum_compare(daos_csum_t *csum, daos_csum_t *csum_src)
+
+static void
+sha1_destroy(struct daos_csummer *obj)
 {
+	D_FREE(obj->dcs_ctx);
+}
 
-#if defined(__x86_64__)
-	return (csum->dc_csum == csum_src->dc_csum) &&
-		!(memcmp(csum->dc_buf, csum_src->dc_buf,
-			 csum_dict[csum->dc_csum].cs_size));
-#else
-	size_t	hash_size;
+static int
+sha1_update(struct daos_csummer *obj,
+		 uint8_t *buf, size_t buf_len)
+{
+	struct mh_sha1_ctx *ctx = obj->dcs_ctx;
 
-	hash_size = mchecksum_get_size(csum_src->dc_csum);
-	mchecksum_get(csum_src->dc_csum, &csum_src->dc_buf, hash_size,
-		      MCHECKSUM_FINALIZE);
-	mchecksum_get(csum->dc_csum, &csum->dc_buf, hash_size,
-		      MCHECKSUM_FINALIZE);
-	return !(strncmp(csum_src->dc_buf, csum->dc_buf, hash_size));
-#endif
+	return mh_sha1_update(ctx, buf, buf_len);
 }
 
 static int
-daos_csum_update(daos_csum_t *csum, const void *buf,
-		 uint64_t len)
-{
-#if defined(__x86_64__)
-	switch (csum->dc_csum) {
-	case DAOS_CS_CRC64:
-	{
-		uint64_t *cur_crc64;
-
-		cur_crc64 = (uint64_t *)csum->dc_buf;
-		*cur_crc64 = crc64_ecma_refl(*cur_crc64,
-					     (const unsigned char *)buf,
-					     len);
+sha1_finish(struct daos_csummer *obj)
+{
+	struct mh_sha1_ctx *ctx = obj->dcs_ctx;
+
+	return mh_sha1_finalize(ctx, obj->dcs_csum_buf);
+}
+
+struct csum_ft sha1_algo = {
+		.update = sha1_update,
+		.init = sha1_init,
+		.destroy = sha1_destroy,
+		.finish = sha1_finish,
+		.csum_len = SHA1_DIGEST_WORDS * 4,
+		.name = "sha1"
+	};
+
+/** ------------------------------------------------------------- */
+static char *csum_unknown_name = "unknown checksum type";
+
+struct csum_ft *
+daos_csum_type2algo(enum DAOS_CSUM_TYPE type)
+{
+	struct csum_ft *result = NULL;
+
+	switch (type) {
+	case CSUM_TYPE_ISAL_CRC16_T10DIF:
+		result = &crc16_algo;
 		break;
-	}
-	case DAOS_CS_CRC32:
-	{
-		uint32_t *cur_crc32;
-
-		cur_crc32 = (uint32_t *)csum->dc_buf;
-		*cur_crc32 = crc32_iscsi((unsigned char *)buf,
-					 (int)len,
-					 *cur_crc32);
+	case CSUM_TYPE_ISAL_CRC32_ISCSI:
+		result = &crc32_algo;
+		break;
+	case CSUM_TYPE_ISAL_CRC64_REFL:
+		result = &crc64_algo;
+		break;
+	case CSUM_TYPE_ISAL_SHA1:
+		result = &sha1_algo;
+		break;
+	case CSUM_TYPE_UNKNOWN:
+	case CSUM_TYPE_END:
 		break;
 	}
-	default:
-		D_ERROR("Unknown checksum type\n");
-		return -DER_NOSYS;
-	}
+	if (result && result->type == CSUM_TYPE_UNKNOWN)
+		result->type = type;
+	return result;
+}
+
+/**
+ * struct daos_csummer functions
+ */
+
+int
+daos_csummer_init(struct daos_csummer **obj, struct csum_ft *ft,
+		      size_t chunk_bytes)
+{
+	if (!ft)
+		return DER_INVAL;
```

Comment:
In general DAOS errors are negated....so -DER_INVAL

Reply:
Fixed